woocommerce/woocommerce-ios
# [Order form] Extract order specific logic from product view models into separate view model for `CollapsibleProductCard`/`CollapsibleProductRowCard`

Closes: #11382
Closes: #11378
⚠️ Based on https://github.com/woocommerce/woocommerce-ios/pull/11374


🗒️ Sorry about the large diffs, about 400-500 diffs were on moving/updating the unit tests from previous view models from https://github.com/woocommerce/woocommerce-ios/pull/11383/commits/5eafa8ee656493730f3d46263b5dbc3ef586a33a 🙇🏻‍♀️

## Why


After the last refactoring PR https://github.com/woocommerce/woocommerce-ios/pull/11370, a bunch of logic/properties were moved from `ProductRowViewModel` to `ProductWithQuantityStepperViewModel`. However, there are still a lot of properties in `ProductRowViewModel` and `ProductWithQuantityStepperViewModel` that are only relevant in the order form. With a new view model for `CollapsibleProductCard` and `CollapsibleProductRowCard` in this PR, a bunch of the properties can be moved to the view model now. To minimize the number of diffs, only the straightforward properties were extracted as listed below. Analytics tracking logic was also moved from `ProductRowViewModel` to the new view model, as these events are only triggered from the order form.

## How

2 new view models were created: `CollapsibleProductCardViewModel` and `CollapsibleProductRowCardViewModel`. `CollapsibleProductCardViewModel` is the top-level view model for `CollapsibleProductCard`, and `CollapsibleProductRowCardViewModel` has most of the properties/logic for the product row. `CollapsibleProductCardViewModel` has the main product row and optional child product rows.

I tried not including `ProductRowViewModel` in `CollapsibleProductRowCardViewModel` since only a subset of properties are used in the order product card, but it resulted in a lot of changes and some order-only properties are still hard to extract because of some private properties. I decided to leave out some properties marked with TODO comments.

- `canChangeQuantity`, `isReadOnly` were moved from `ProductWithQuantityStepperViewModel` to `CollapsibleProductRowCardViewModel`
- `childProductRows` was moved from `ProductWithQuantityStepperViewModel` to `CollapsibleProductCardViewModel`
- `hasDiscount`, `totalPriceAfterDiscountLabel`, `discountLabel`, and `hasParentProduct` were moved from `ProductRowViewModel` to `CollapsibleProductRowCardViewModel`
- `trackAddDiscountTapped` and `trackEditDiscountTapped` functions were moved from `ProductRowViewModel` to `CollapsibleProductRowCardViewModel`

`SimplifiedProductRow` struct was also removed with its subviews moved to `CollapsibleProductRowCard`.

## Testing instructions


No changes on the order form UX are expected. The following steps might not cover all the possible cases, please feel free to diverge a bit and share any issues!

- Go to the Orders tab
- Tap on `+` to create an order
- Tap `Add Products` --> the product rows in the product selector should behave the same as before
- Search/select at least one product and tap the bottom CTA to confirm the selection --> the selected product(s) should show quantity as 1
- Expand any product card
- Tap to increment the product quantity --> the quantity label in the price in the product card should also be updated
- After the payments section is synced, tap `Create` to create the order --> the created order should have the recently set quantity for each product
- Tap `Edit` to edit the order --> the quantity for each product should match what was set previously
- Tap to increment the product quantity --> after syncing, the quantity label in the price in the product card should be updated
- Tap to decrease the product quantity to 0 --> the product should be removed
- Tap `Done` --> the updated order should have the recently set quantity for each product

---
